misc-scripts/xref_mapping/XrefParser/RefSeqGPFFParser.pm

```diff
@@ -17,340 +17,341 @@ limitations under the License.
-      foreach my $cb (@coded_by){
-	$xref->{PAIR} = $cb;
-      }
-      if (!defined $xref->{PAIR}) {
-        $xref->{PAIR} = $refseq_pair;
-      }
+# returns the source id for a source name, requires $self->{source_ids} to have been populated
+sub source_id_from_name {
+  my ($self, $name) = @_;
 
-      foreach my $pi (@protein_id){
-	$xref->{PROTEIN} = $pi;
-      }
+  my $source_id;
 
-      foreach my $ll (@EntrezGeneIDline) {
-	my %dep;
-        #my $entrez_source = $dependent_sources{EntrezGene} || die( 'No source for EntrezGene!' );
-        #my $wiki_source = $dependent_sources{WikiGene} || die( 'No source for WikiGene!' );
-        if (defined $entrez{$ll}) {
-          $dep{SOURCE_ID} = $entrez_source_id;
-	  $dep{LINKAGE_SOURCE_ID} = $source_id;
-	  $dep{ACCESSION} = $ll;
-          $dep{LABEL} = $entrez{$ll};
-	  push @{$xref->{DEPENDENT_XREFS}}, \%dep;
-
-  	  my %dep2;
-	  $dep2{SOURCE_ID} = $wiki_source_id;
-	  $dep2{LINKAGE_SOURCE_ID} = $source_id;
-	  $dep2{ACCESSION} = $ll;
-          $dep2{LABEL} = $entrez{$ll};
-	  push @{$xref->{DEPENDENT_XREFS}}, \%dep2;
-
-          # Add xrefs for RefSeq mRNA as well where available
-          $refseq_pair =~ s/\.[0-9]*// if $refseq_pair;
-          if (defined $refseq_pair) {
-            if ($refseq_ids->{$refseq_pair}) {
-              foreach my $refseq_id (@{ $refseq_ids->{$refseq_pair} }) {
-                foreach my $entrez_id (@{ $entrez_ids->{$ll} }) {
-                  $add_dependent_xref_sth->execute($refseq_id, $entrez_id);
-                }
-                foreach my $wiki_id (@{ $wiki_ids->{$ll} }) {
-                  $add_dependent_xref_sth->execute($refseq_id, $wiki_id);
-                }
-              }
-            }
-          }
-        }
-      }
+  if ( exists $self->{source_ids}->{$name} ) {
+    $source_id = $self->{source_ids}->{$name};
+  } elsif ( $self->{verbose} ) {
+    warn "WARNING: can't get source ID for name '$name'\n";
+  }
 
-      # Don't add SGD Xrefs, as they are mapped directly from SGD ftp site
+  return $source_id;
+}
+
+# returns the source id for a RefSeq accession, requires $self->{source_ids} to have been populated
+sub source_id_from_acc {
+  my ($self, $acc) = @_;
+
+  my $source_id;
+  my $prefix = substr($acc, 0, 2);
 
-      # Refseq's do not tell whether the mim is for the gene of morbid so ignore for now.
-      return $xref;
+  if ( exists $REFSEQ_SOURCES->{$prefix} ) {
+    $source_id = $self->source_id_from_name( $REFSEQ_SOURCES->{$prefix} );
+  } elsif ( $self->{verbose} ) {
+    warn "WARNING: can't get source ID for accession '$acc'\n";
   }
+
+  return $source_id;
 }
 
-# --------------------------------------------------------------------------------
+# get type from filename path. this includes the source name and that's enough to extract it
+sub type_from_file {
+  my ($self, $file) = @_;
+
+  my ($type) = $file =~ /RefSeq_(\w+)\//x;
+
+  warn "WARNING: Could not work out sequence type for '$file'\n" unless $type;
```

Comment:
As far as I can see, after your refactor the only place `$type` is used is in
```
    SEQUENCE_TYPE => $params->{type},
```
(apart from `WormbaseCElegansRefSeqGPFFParser`). 

If that's the case, it can only be a few things that are later accepted, I think in the sequence match part of the pipeline. You mentioned restricting it to just mrna / protein - do that! Maybe even with a big comment explaining the dependency and what this is tied to, referring to constants in the module where the sequence match code is, or something like that.

Reply:
type is now always based on the name of the source.
That is RefSeq_peptide and RefSeq_dna that translate into types peptide and dna.